CR-2760: [CAR] [GCP] Ensure Google compute firewall ingress does not allow unrestricted ssh access

Imanuel-Mizrahi opened the pull request.

Tomerva commented on cloudrail/knowledge/rules/gcp/context_aware/public_access_vpc_port_rule.py:
```diff
@@ -0,0 +1,87 @@
+                    issues.append(
+                        Issue(
+                            f"The {network_entity.get_type()} `{network_entity.get_friendly_name()}` "
+                            f"with one of the public IP addresses `{', ' .join(map(str,public_ip_addresses))}` "
```
why did you need to use map? (also youre missing a space after str,)

Tomerva commented on tests/knowledge/context/gcp/test_compute_forwarding_rule.py:
```diff
@@ -0,0 +1,20 @@
+        if compute.origin == EntityOrigin.LIVE_ENV:
+            self.assertEqual(compute.target, 'https://www.googleapis.com/compute/v1/projects/dev-for-tests/regions/us-west1/targetPools/default-tp')
+        else:
+            self.assertEqual(compute.target, 'google_compute_target_pool.default.self_link')
```
i think None would be better if you cant build the target manually.. having the tf address is probably useless

Imanuel-Mizrahi replied:
No, it is still needed for relations assigner, and it works.

Imanuel-Mizrahi commented on cloudrail/knowledge/context/environment_context/common_component_builder.py:
```diff
@@ -91,3 +91,8 @@ def extract_attribute_from_file_path(path: str, strings_to_remove: list):
         attribute = attribute.replace(string, '')
     return attribute.replace('.json', '')
+
+def extract_name_from_gcp_link(gcp_link: str) -> str:
```
That it would be a TF address

Imanuel-Mizrahi replied:
true, could be the entire function. Tested.

Imanuel-Mizrahi commented on cloudrail/knowledge/context/gcp/resources_builders/scanner/compute_forwarding_rule_builder.py:
```diff
@@ -0,0 +1,14 @@
+    def do_build(self, attributes: dict) -> GcpComputeForwardingRule:
+        return GcpComputeForwardingRule(name=attributes['name'],
+                                        target=attributes['target'],
+                                        port_range=PortSet(attributes['portRange'].split('-')))
```
because the format of "attributes['portRange']" is a string representing range, like this: "22-22".
PortSet does not know how to handle this, so when I split it - this works.

Tomerva replied:
it knows how to handle it, moreover, what you did is not right as for example for the range 22-25 you will get a port set with 22 and 25

![image](https://user-images.githubusercontent.com/10116471/144065400-db899f3a-f334-4ce9-80c8-fa2f75684167.png)

Imanuel-Mizrahi replied:
Removed the split, I remember the function crashed for me, but cannot figure out exactly why I guess

Imanuel-Mizrahi commented on cloudrail/knowledge/context/gcp/gcp_connection_builder.py:
```diff
@@ -1,12 +1,91 @@
+                                    self._filter_connections_by_action(denied_connections_by_priority, allowed_connections_by_priority))
+        return connections_set
+
+    def _filter_connections_by_priority(self, connections_list: Set[GcpConnection]) -> set:
```
I have completely changes the connections builder, you should look again once done.

Imanuel-Mizrahi commented on cloudrail/knowledge/context/gcp/gcp_relations_assigner.py:
```diff
@@ -10,6 +18,56 @@ def __init__(self, ctx: GcpEnvironmentContext = None):
+    @staticmethod
+    def _assign_targets_to_forward_rule(forward_rule: GcpComputeForwardingRule, target_pools: List[GcpComputeTargetPool]):
+        def get_target():
+            target = next((target for target in target_pools if target.self_link == forward_rule.target), None)
```
yes, it does.

Imanuel-Mizrahi commented on cloudrail/knowledge/context/gcp/resources_builders/terraform/compute_firewall_builder.py:
```diff
@@ -35,16 +36,14 @@ def do_build(self, attributes: dict) -> GcpComputeFirewall:
 
     @staticmethod
-    def get_action_block_data(attributes: dict) -> dict:
+    def get_action_block_data(attributes: dict, function: Callable) -> dict:
```
changed to: "get_known_value_func"

Imanuel-Mizrahi commented on cloudrail/knowledge/context/gcp/resources_builders/terraform/compute_forwarding_rule_builder.py:
```diff
@@ -0,0 +1,22 @@
+    def do_build(self, attributes: dict) -> GcpComputeForwardingRule:
+        port_range = None
+        if port_range := self._get_known_value(attributes, 'port_range'):
+            if ',' in port_range:
```
Not sure exactly, we might need to discuss this.

Tomerva replied:
why not? just need to split comma, remove trailing whitespaces, and then for each item convert it using the existing parse-from-string way

Imanuel-Mizrahi replied:
Turns out that TF docs were confusing, there will be no scenario needed for this change.

Tomerva commented on cloudrail/knowledge/context/gcp/gcp_connection_builder.py:
```diff
@@ -1,12 +1,61 @@
+    def _filter_conns_by_priority_and_action(connections_list: Set[GcpConnection]) -> Set[GcpConnection]:
+        connections_set: Set[GcpConnection] = set()
+        for connection in connections_list:
+            if not any(connection.__eq__(list_con) for list_con in connections_set):
```
i dont think this algorithm is right..
first of all, youre not considering partial cidr blocks overlap. for example
Rule #1: Cidr: 1.1.1.1/32, DENY
Rule #2: Cidr: 1.1.1.1/16, ALLOW

with this rule set, 1.1.1.1/16 is allowed except for ip 1.1.1.1

second, you are only filtering rules that have some other rule with same priority, thats not how this works.
read https://cloud.google.com/vpc/docs/firewalls#priority_order_for_firewall_rules 
the documentation on the start talks only about same-priority rules, but look at the example..

Tomerva commented on cloudrail/knowledge/context/gcp/resources/networking_config/network_entity.py:
```diff
@@ -0,0 +1,80 @@
+                     self.priority,
+                     self.firewall))
+
+    def __eq__(self, other) -> bool:
```
i think theres a slight difference in the meaning of the function here..
you implement equality but it feels like this i something more specific - it checks that the target of the connection is the same, as you ignore priority and firewall_action.

so instead of using implementing `__eq__` i would replace it with a function like `compare_target(self, other)` or something like that

Tomerva commented on cloudrail/knowledge/context/gcp/resources/networking_config/network_entity.py:
```diff
@@ -0,0 +1,80 @@
+        ConnectionInstance.__init__(self)
+        self.network_interfaces: List[GcpNetworkInterface] = network_interfaces
+        self.forwarding_rules: List[GcpComputeForwardingRule] = []
+        self.inbound_connections: Set[GcpConnection] = set()
```
arent these already part of ConnectionInstance ?

Tomerva commented on cloudrail/knowledge/context/gcp/resources_builders/terraform/compute_forwarding_rule_builder.py:
```diff
@@ -0,0 +1,22 @@
+            if ',' in port_range:
+                port_range = PortSet(port_range.split(','))
+            else:
+                port_range = PortSet([port_range, port_range])
```
why is port_range specified twice?

Tomerva commented on cloudrail/knowledge/rules/gcp/context_aware/public_access_vpc_port_rule.py:
```diff
@@ -0,0 +1,100 @@
+
+    @staticmethod
+    def _is_effected_ip_protocol(conn: GcpConnection, ip_protocol: IpProtocol) -> bool:
+        return conn.connection_property.ip_protocol_type in (IpProtocol.ALL, ip_protocol)
```
IpProtocol already implements methods for comparisons, use them

Tomerva commented on cloudrail/knowledge/rules/gcp/context_aware/public_access_vpc_port_rule.py:
```diff
@@ -0,0 +1,100 @@
+
+    @staticmethod
+    def _is_public_inbound_conn(conn: GcpConnection) -> bool:
+        return conn.connection_type == ConnectionType.PUBLIC \
```
fix indentation

Tomerva commented on cloudrail/knowledge/rules/gcp/context_aware/public_access_vpc_port_rule.py:
```diff
@@ -0,0 +1,100 @@
+    def _is_public_inbound_conn(conn: GcpConnection) -> bool:
+        return conn.connection_type == ConnectionType.PUBLIC \
+            and conn.connection_direction_type == ConnectionDirectionType.INBOUND \
+                and conn.connection_property.cidr_block == '0.0.0.0/0'
```
this is a very specific case of a public connection.. you should use the function that checks if a given cidr is on the public domain

Tomerva commented on cloudrail/knowledge/rules/gcp/context_aware/public_access_vpc_port_rule.py:
```diff
@@ -0,0 +1,100 @@
+    def conns_allowed_by_firewall_on_port(self, network_resource: NetworkEntity, port: int) -> GcpConnection:
+        return [conn for conn in network_resource.inbound_connections
+                if conn.firewall_action == FirewallRuleAction.ALLOW
+                and self._is_effected_ip_protocol(conn, IpProtocol('TCP'))
```
why is it on TCP? what about UDP?

Tomerva commented on cloudrail/knowledge/rules/gcp/context_aware/public_access_vpc_port_rule.py:
```diff
@@ -0,0 +1,100 @@
+            if connections:
+                forwarding_rules = self.conns_forwarding_on_port(network_entity, self.port)
+                public_ip_addresses = network_entity.public_ip_addresses
+                for connection in connections:
```
the logic here is incorrect - you will always add issue as per connections, regardless if the connections are private or public

Imanuel-Mizrahi commented on cloudrail/knowledge/context/gcp/gcp_relations_assigner.py:
```diff
@@ -10,6 +18,59 @@ def __init__(self, ctx: GcpEnvironmentContext = None):
+        implied_firewalls = self.pseudo_builder.get_implied_firewalls()
+        for vpc in vpcs:
+            for firewall in implied_firewalls:
+                firewall.network = vpc.self_link
```
True, but this is intentional - according to the [docs](https://cloud.google.com/vpc/docs/firewalls#default_firewall_rules), ALL VPC's will have this default FW rules.
This doc reference appear just before this function in the code.
So, I am adding it to each VPC, just replacing the network attribute of the FW, with the VPC's network attribute.

Imanuel-Mizrahi commented on cloudrail/knowledge/context/gcp/gcp_relations_assigner.py:
```diff
@@ -10,6 +18,59 @@ def __init__(self, ctx: GcpEnvironmentContext = None):
+    @staticmethod
+    def _assign_networking_info_to_network_interfaces(instance: GcpComputeInstance, vpcs: List[GcpComputeNetwork]):
+        def get_instance_vpcs():
+            instance_vpcs = [vpc for vpc in vpcs if any(extract_name_from_gcp_link(interface.network) == vpc.name for interface in instance.network_interfaces)]
```
No, actually, it's not possible.
If you try to create 2 NIC's in the same VPC, it will throw an error:
"Networks must be distinct for NICs attached to a VM"
Anyway, I do not think there is an issue with the function, right?

Imanuel-Mizrahi commented on cloudrail/knowledge/context/gcp/gcp_relations_assigner.py:
```diff
@@ -10,6 +18,59 @@ def __init__(self, ctx: GcpEnvironmentContext = None):
+            for nic in instance.network_interfaces:
+                vpc_network = next((vpc for vpc in instance_vpcs if extract_name_from_gcp_link(nic.network) == vpc.name), None)
+                nic.vpc_network = vpc_network
+                nic.firewalls = vpc_network.firewalls
```
Do you want me to assign it in the base class itself, like that:
"self.firewalls = self.vpc_network.firewalls"?

Imanuel-Mizrahi commented on cloudrail/knowledge/context/gcp/gcp_relations_assigner.py:
```diff
@@ -10,6 +18,59 @@ def __init__(self, ctx: GcpEnvironmentContext = None):
+
+    @staticmethod
+    def _assign_forward_rule_to_instance(compute_instance: GcpComputeInstance, forwarding_rules: List[GcpComputeForwardingRule]):
+        if any(rule.target_pool for rule in forwarding_rules):
```
modified to:
`    @staticmethod
    def _assign_forward_rule_to_instance(compute_instance: GcpComputeInstance, forwarding_rules: List[GcpComputeForwardingRule]):
        def get_forwarding_rules():
            forwarding_rules_list = [rule for rule in forwarding_rules if rule.target_pool and compute_instance.self_link in rule.target_pool.instances]
            return forwarding_rules_list

        compute_instance.forwarding_rules = (ResourceInvalidator.get_by_logic(get_forwarding_rules, False))`

Imanuel-Mizrahi commented:
@Tomerva , I added a new file: gcp_connections_evaluator.py
The approach I took now is as follows:
The GCP rules evaluation depends on the order of the rules, first one to match - will be enforced (this is actually the meaning of the priority eventually).
So, the key in the connection building, and evaluation is to sort the rules properly.
Now the connections will be sorted by priority, and if the rules has the same priority, the DENY rules will be on top.
I changed the GcpConnections to be a list and not set, so we will maintain the order of the rules, and the loop iterations inside the gcp_connections_evaluator will be done by the order of the rules.
In this file (gcp_connections_evaluator), we will have all the evaluation logics needed per demand of product.
Which by the way, all your comments regarding the rule logic, should be checked with product, as this is the ticket requirenment - we should probably create a list of questions and discuss with them.
Thanks!

Viperain commented on cloudrail/knowledge/context/gcp/resources/networking_config/network_entity.py:
```diff
@@ -0,0 +1,80 @@
+    priority: int
+    firewall: GcpComputeFirewall
+
+    def __hash__(self) -> int:
```
do not override hash or eq methods. as tomer wrote you can imeplent compare_target if you need different compare

Imanuel-Mizrahi commented on cloudrail/knowledge/context/gcp/resources/compute/gcp_compute_firewall.py:
```diff
@@ -82,3 +87,12 @@ def get_type(self, is_plural: bool = False) -> str:
                 'source_ranges': self.source_ranges}
+
+    @property
+    def firewall_ip_ranges(self) -> set:
```
Eventually in the gcp_connection_builder, this is being separated by the connection direction
